locator/topology.cc

```diff
@@ -200,18 +202,25 @@ int topology::compare_endpoints(const inet_address& address, const inet_address&
         sstring a1_rack = get_rack(a1);
         sstring a2_rack = get_rack(a2);
 
-        if (address_rack == a1_rack && address_rack != a2_rack) {
-            return -1;
-        }
-
-        if (address_rack == a2_rack && address_rack != a1_rack) {
+        // If one of the Nodes IS in the same Rack as the Node
+        // we are comparing to and the other one
+        // IS NOT - then return the appropriate result.
+        if (address_rack == a1_rack) {
+            if (address_rack != a2_rack) {
+                return -1;
+            }
+        } else if (address_rack == a2_rack) {
             return 1;
         }
+        // We don't differentiate between Nodes if all Nodes belong to
+        // the same Rack or to different Racks (in the same Data Center),
+        // thus make them equal.
+    } else if (address_datacenter == a2_datacenter) {
+        return 1;
+    } else {
+        // We don't differentiate between Nodes if all Nodes belong to different
+        // Data Centers, thus make them equal.
     }
-    //
-    // We don't differentiate between Nodes if all Nodes belong to different
-    // Data Centers, thus make them equal.
-    //
```

Comment:
Idea (needs node_ptr first): give each node a metric `(dc_seq << 48) | (rack_seq << 32) | node_seq`. The actual sequence numbers don't matter.

Then, this function is abs(n1.metric - n.metric) <=> (n2.metric - n.metric).

Reply:
It won't be exactly backward compatible for nodes in different racks in same dc, but I think it's work just as well

Reply:
or even nodes in the same rack, for both cases we return 0 today, but 0 here stands for "don't care" as as I understood from the comments.

Reply:
I think this would be a nice follow up over https://github.com/scylladb/scylladb/pull/11987
where we can keep and get the metric in the node object.
In the current codebase we can extend topology or endpoint_dc_rack to compute and keep the metric, but I'd rather spend the effort in the nodes model.

Reply:
Since the metric is local, we could assign the same value for remote dc/demote racks.

`metric = !is_same_node | (!is_same_rack << 1) | (!is_same_dc << 1)`

7 = very far, 3 = local dc, 1 = same rack, 0 = looking in mirror

Reply:
and, of course, it could be precalculated.